src/amazon/dsstne/engine/NNNetwork.cpp

```diff
@@ -1900,111 +1905,153 @@ bool NNNetwork::SaveNetCDF(const string& fname)
     return bResult;
 }
 
-vector<string> NNNetwork::GetLayers()
+vector<string> NNNetwork::GetLayers() const
 {
     vector<string> vResult;
     for (auto l: _vLayer)
+    {
         vResult.push_back(l->_name);
+    }
+    
     return vResult;
 }
 
 
-NNFloat* NNNetwork::GetUnitBuffer(const string& layer)
+const NNFloat* NNNetwork::GetUnitBuffer(const string& layer) const
 {
-    NNLayer* pLayer         = _mLayer[layer];
-    if (pLayer == NULL)
+    const auto itr = _mLayer.find(layer);
+    if (itr == _mLayer.end())
     {
         if (getGpu()._id == 0)
+        {
             printf("NNNetwork::GetUnitBuffer: Unknown layer %s.\n", layer.c_str());
+        }
+        
         return NULL;
     }
 
-    return pLayer->GetUnitBuffer();
+    return itr->second->GetUnitBuffer();
 }
 
-NNFloat* NNNetwork::GetDeltaBuffer(const string& layer)
+const NNFloat* NNNetwork::GetDeltaBuffer(const string& layer) const
 {
-    NNLayer* pLayer         = _mLayer[layer];
-    if (pLayer == NULL)
+    const auto itr = _mLayer.find(layer);
+    if (itr == _mLayer.end())
     {
         if (getGpu()._id == 0)
+        {
             printf("NNNetwork::GetDeltaBuffer: Unknown layer %s.\n", layer.c_str());
+        }
+
         return NULL;
     }
 
-    return pLayer->GetDeltaBuffer();
+    return itr->second->GetDeltaBuffer();
 }
 
-uint64_t NNNetwork::GetBufferSize(const string& layer)
+uint64_t NNNetwork::GetBufferSize(const string& layer) const
 {
-    NNLayer* pLayer         = _mLayer[layer];
-    if (pLayer == NULL)
+    const auto itr = _mLayer.find(layer);
+    if (itr == _mLayer.end())
     {
         if (getGpu()._id == 0)
+        {
             printf("NNNetwork::GetDeltaBuffer: Unknown layer %s.\n", layer.c_str());
+        }
+
         return 0;
     }
 
-    return pLayer->GetBufferSize();
+    return itr->second->GetBufferSize();
 }
 
-NNWeight* NNNetwork::GetWeight(const string& inputLayer, const string& outputLayer)
+const NNWeight* NNNetwork::GetWeight(const string& inputLayer, const string& outputLayer) const
 {
-    NNLayer* pInputLayer    = _mLayer[inputLayer];
-    NNLayer* pOutputLayer   = _mLayer[outputLayer];
-
-    // Validate parameters
-    if (pInputLayer == NULL)
+    const auto inputLayerItr = _mLayer.find(inputLayer);
+    if (inputLayerItr == _mLayer.end())
     {
         if (getGpu()._id == 0)
+        {
             printf("NNNetwork::GetWeight: Unknown input layer %s.\n", inputLayer.c_str());
+        }
+
         return NULL;
     }
-    if (pOutputLayer == NULL)
+    
+    const auto outputLayerItr = _mLayer.find(inputLayer);
```

Comment:
it is a bug: it should be
const auto outputLayerItr = _mLayer.find(outputLayer);